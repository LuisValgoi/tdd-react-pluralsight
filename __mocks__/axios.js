const getMock = jest.fn().mockResolvedValue({
  data: {
    shops: [
      {
        name: "Portland",
        image: "portland",
        address: "Avenue Portland, 123",
      },
      {
        name: "Astoria",
        image: "astoria",
        address: "Avenue Astoria, 123",
      },
      {
        name: "",
        image: "none",
        address: "N/A",
      },
    ],
  },
});

const axiosMock = {
  get: getMock,
};

export default axiosMock;
