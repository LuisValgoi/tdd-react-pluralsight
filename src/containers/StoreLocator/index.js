import React, { useEffect, useState } from "react";
import Header from "../../components/Header";
import Button from "../../components/Button";
import Map from "../../components/Map";
import { fetchShops } from "../../services/shops";

export default function StoreLocator() {
  const [shops, setShops] = useState([]);
  const [currentMap, setCurrentMap] = useState({});

  const handleButtonClick = (image) => {
    const foundShop = shops.find((location) => image === location.image);
    setCurrentMap(foundShop);
  };

  useEffect(() => {
    async function loadData() {
      const { data } = await fetchShops();
      setShops(data);
      setCurrentMap(data[data.length - 1]);
    }
    loadData();
  }, []);

  return (
    <div data-testid="store-locator">
      <Header />
      <div data-testid="store-locator-buttons">
        {shops.map((store, index) => {
          return <Button onClick={() => handleButtonClick(store.image)} key={index} name={store.name} />;
        })}
      </div>
      {shops && <Map image={currentMap.image} locationName={currentMap.name} />}
    </div>
  );
}
