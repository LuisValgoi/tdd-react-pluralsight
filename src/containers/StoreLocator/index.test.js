import { act, fireEvent, render, screen, waitFor } from "@testing-library/react";
import StoreLocator from "./index";
import axios from "axios";

function mockCall() {
  axios.get.mockImplementation(() =>
    Promise.resolve({
      data: [
        {
          name: "Portland",
          image: "portland",
          address: "Avenue Portland, 123",
        },
        {
          name: "AstoriaUAEHAUHEAUHE",
          image: "astoria",
          address: "Avenue Astoria, 123",
        },
        {
          name: "",
          image: "none",
          address: "N/A",
        },
      ],
    })
  );
}

jest.mock("axios");

beforeEach(() => {
  mockCall();
});

describe("StoreLocator Suite", () => {
  test("renders correctly", () => {
    const { container } = render(<StoreLocator />);

    expect(container).toMatchSnapshot();
  });

  test("renders StoreLocator without crashing", async () => {
    await waitFor(() => {
      render(<StoreLocator />);

      const storeLocatorElement = screen.getByTestId("store-locator");

      expect(storeLocatorElement).toBeInTheDocument();
    });
  });

  test("renders Header", async () => {
    await waitFor(() => {
      render(<StoreLocator />);

      const headerElement = screen.getByTestId("header");

      expect(headerElement).toBeInTheDocument();
    });
  });

  test("renders no buttons right away", async () => {
    await waitFor(() => {
      render(<StoreLocator />);

      const storeLocatorButtonsElement = screen.getByTestId("store-locator-buttons");

      expect(storeLocatorButtonsElement).toBeInTheDocument();
      expect(storeLocatorButtonsElement.children).toHaveLength(0);
    });
  });
});

describe("StoreLocator Suite - chooseMap", () => {
  test("updates the map to the portland image once pressed on portland button", async () => {
    await waitFor(() => {
      render(<StoreLocator />);

      const buttonElement = screen.getByRole("button", { name: /portland/i });
      fireEvent.click(buttonElement);

      const imgElement = screen.getByText(/portland/i);
      expect(imgElement).toBeInTheDocument();
    });
  });
});
