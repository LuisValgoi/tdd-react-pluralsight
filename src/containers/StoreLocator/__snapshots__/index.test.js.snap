// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreLocator Suite renders correctly 1`] = `
<div>
  <div
    data-testid="store-locator"
  >
    <header
      data-testid="header"
    >
      <img
        alt="Wired Brain Logo"
        src="/images/wired-brain-coffee-logo.png"
      />
    </header>
    <div
      data-testid="store-locator-buttons"
    />
    <div
      class="map-box"
      data-testid="map"
    >
      <img
        src="/images/none.png"
      />
    </div>
  </div>
</div>
`;
